{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup, Bot\n",
    "from telegram.ext import Updater, CommandHandler, CallbackQueryHandler, MessageHandler, filters, CallbackContext\n",
    "from telegram.ext import Application, ContextTypes\n",
    "import nest_asyncio\n",
    "nest_asyncio.apply()\n",
    "\n",
    "from database import DataBaseFetch\n",
    "\n",
    "TOKEN = \"6929830229:AAEXbYO97fey0HwecRuIPFTLXYT-WxzgigI\"\n",
    "MANAGER = {\"id\":\"ali_zeighamiyan\"}\n",
    "ROLES = [\"employee\", \"stockman\"]\n",
    "WORKERS = []\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_fetcher = DataBaseFetch(db_name=\"user_data.db\")\n",
    "[db_fetcher.add_role(role) for role in ROLES]\n",
    "db_fetcher.add_worker(\"ali\", \"z\")\n",
    "db_fetcher.assign_role_to_worker(\"ali\", ROLES[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "ADD_WORKER = {\"text\":\"Add Worker\", \"callback_data\":\"add_worker\"}\n",
    "ADD_ANOTHER_WORKER =  {\"text\":\"Add Another Worker\", \"callback_data\":\"add_worker\"}\n",
    "VIEW_WORKER = {\"text\":\"View Workers\", \"callback_data\":\"view_workers\"}\n",
    "BACK = {\"text\":\"Back\", \"callback_data\":\"back\"}\n",
    "ASSIGN_ROLE = {\"text\":\"Assign Role\", \"callback_data\":\"select_role\"}\n",
    "DELETE_WORKER = {\"text\":\"Delete Worker\", \"callback_data\":\"deleteworker\"}\n",
    "\n",
    "MANAGER_START_MENU = [ADD_WORKER, VIEW_WORKER]\n",
    "WORKER_START_MENU = [VIEW_WORKER]\n",
    "EDIT_WORKER_MENU = [ASSIGN_ROLE, DELETE_WORKER, BACK]\n",
    "ADD_WORKER_NAME_MENU = [BACK]\n",
    "ADD_WORKER_ID_MENU = [BACK, ADD_ANOTHER_WORKER]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ButtonMaker:\n",
    "    def __init__(self) -> None:\n",
    "        self.keyboard = []\n",
    "    def build_keyboard(self, text_callback_list:list[dict]):\n",
    "        for text_callback_data in text_callback_list:\n",
    "            self.keyboard.append([InlineKeyboardButton(text_callback_data[\"text\"], callback_data=text_callback_data[\"callback_data\"])])\n",
    "        return self\n",
    "    def get_markup(self):\n",
    "        reply_markup = InlineKeyboardMarkup(self.keyboard)\n",
    "        self.keyboard = []\n",
    "        return reply_markup\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "button_maker = ButtonMaker()\n",
    "buttons = {\n",
    "           \"MANAGER_START_MENU\" :  button_maker.build_keyboard(MANAGER_START_MENU).get_markup(),\n",
    "           \"WORKER_START_MENU\" : button_maker.build_keyboard(WORKER_START_MENU).get_markup(),\n",
    "           \"ADD_WORKER_NAME_MENU\" : button_maker.build_keyboard(ADD_WORKER_NAME_MENU).get_markup(),\n",
    "           \"ADD_WORKER_ID_MENU\" : button_maker.build_keyboard(ADD_WORKER_ID_MENU).get_markup(),\n",
    "           \"EDIT_WORKER_MENU\" : button_maker.build_keyboard(EDIT_WORKER_MENU).get_markup(),\n",
    "           }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_fetcher = DataBaseFetch(db_name=\"user_data.db\")\n",
    "# In-memory storage for users, workers, and roles\n",
    "\n",
    "managers = {\"ali_zeighamiyan\": True}  # Replace with actual manager usernames\n",
    "\n",
    "# def get_worker_buttons_markup(workers:list, msg:str):\n",
    "#     if not workers :\n",
    "#         workers.append((\"No Worker Found!\", ))\n",
    "#     worker_buttons = [InlineKeyboardButton(worker[0], callback_data=f\"worker-{msg}:{worker[0]}\") for worker in workers]\n",
    "#     worker_markup = InlineKeyboardMarkup.from_column(worker_buttons)\n",
    "#     return worker_markup\n",
    "\n",
    "\n",
    "async def edit_worker(query, context):\n",
    "    worker_name = query.data.split(\":\")[1]\n",
    "    context.user_data[\"selected-worker\"] = worker_name\n",
    "    reply_markup = buttons[\"EDIT_WORKER_MENU\"]\n",
    "    await query.edit_message_text(f'Worker:{worker_name}', reply_markup=reply_markup)\n",
    "\n",
    "async def start_menu(query, context) -> None:\n",
    "    \"\"\"Send a message when the command /start is issued.\"\"\"\n",
    "    context.user_data[\"pos-stack\"] = [\"start_menu\"]\n",
    "    username = context.user_data[\"username\"]\n",
    "    if username in managers:\n",
    "        reply_markup = buttons[\"MANAGER_START_MENU\"]\n",
    "    else:\n",
    "        reply_markup = buttons[\"WORKER_START_MENU\"]\n",
    "    await query.edit_message_text('Welcome to the Worker Manager Bot! Choose an option:', reply_markup=reply_markup)\n",
    "    \n",
    "    \n",
    "async def start(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:\n",
    "    \"\"\"Send a message when the command /start is issued.\"\"\"\n",
    "    context.user_data[\"pos-stack\"] = [\"start_menu\"]\n",
    "    username = update.message.from_user.username\n",
    "    if username in managers:\n",
    "        reply_markup = buttons[\"MANAGER_START_MENU\"]\n",
    "    else:\n",
    "        reply_markup = buttons[\"WORKER_START_MENU\"]\n",
    "    await update.message.reply_text('Welcome to the Worker Manager Bot! Choose an option:', reply_markup=reply_markup)\n",
    "    \n",
    "        \n",
    "async def button(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:\n",
    "    \"\"\"Handle button presses for various commands.\"\"\"\n",
    "    query = update.callback_query\n",
    "    await query.answer()\n",
    "\n",
    "    data = query.data\n",
    "    username = query.from_user.username\n",
    "    context.user_data[\"username\"] = username\n",
    "\n",
    "    if data != \"back\":\n",
    "        context.user_data[\"pos-stack\"].append(data)\n",
    "    if data == 'add_worker' and username in managers:\n",
    "        await add_worker(query, context)\n",
    "    elif data == 'select_role' and username in managers:\n",
    "        await select_role(query, context)\n",
    "    elif data == 'view_workers':\n",
    "        await view_workers(query, context)\n",
    "    elif data == \"delete_worker\":\n",
    "        await delete_worker(query, context)\n",
    "        \n",
    "    elif data.startswith(\"handle_role_assignment\"):\n",
    "        await handle_role_assignment(query, context)\n",
    "    \n",
    "    elif data.startswith(\"edit_worker\"):\n",
    "        await edit_worker(query, context)\n",
    "    elif data == 'back':\n",
    "        context.user_data[\"pos-stack\"].pop()\n",
    "        pos_to_go = context.user_data[\"pos-stack\"][-1]\n",
    "        await globals()[pos_to_go](query, context)\n",
    "\n",
    "async def view_workers(query, context):\n",
    "    \"\"\"Display the list of workers.\"\"\"\n",
    "    workers = db_fetcher.get_workers()\n",
    "    if not workers: workers.append((\"No Worker Found!\", ))\n",
    "    button_maker.build_keyboard([{\"text\":worker[0], \"callback_data\":f\"edit_worker:{worker[0]}\"}\n",
    "                                    for worker in workers])\n",
    "    button_maker.build_keyboard([BACK])\n",
    "    worker_markup = button_maker.get_markup()\n",
    "    await query.edit_message_text('Choose From Workers', reply_markup=worker_markup)\n",
    "\n",
    "    # await query.edit_message_text(f\"Workers:\\n{worker_list}\")\n",
    "\n",
    "async def select_role(query, context):\n",
    "    worker_name = context.user_data[\"selected-worker\"]\n",
    "    button_maker.build_keyboard([{\"text\":role, \"callback_data\":f\"handle_role_assignment:{role}\"}\n",
    "                                               for role in ROLES])\n",
    "    role_markup = button_maker.build_keyboard([BACK]).get_markup()\n",
    "    \n",
    "    # role_buttons = [InlineKeyboardButton(role, callback_data=f\"role-assign:{role}\") for role in ROLES]\n",
    "    # role_markup = InlineKeyboardMarkup.from_column(role_buttons)\n",
    "    await query.edit_message_text(f'Select a role to assign to {worker_name}:', reply_markup=role_markup)\n",
    "\n",
    "async def handle_role_assignment(query, context):\n",
    "    role_name = query.data.split(\":\")[1]\n",
    "    worker_name = context.user_data.get('selected-worker')\n",
    "    if worker_name:\n",
    "        db_fetcher.assign_role_to_worker(worker_name=worker_name, role_name=role_name)\n",
    "        await query.edit_message_text(f'Assigned role {role_name} to worker {worker_name}.')\n",
    "        \n",
    "\n",
    "async def handle_message(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:\n",
    "    \"\"\"Handle incoming messages.\"\"\"\n",
    "    # user_id = update.message.from_user.id\n",
    "\n",
    "    if context.user_data.get('adding_worker'):\n",
    "        state = context.user_data.get(\"state\")\n",
    "        \n",
    "        if state == \"AddWorkerName\":\n",
    "            context.user_data[\"worker_detail\"] = {}\n",
    "            worker_name = update.message.text\n",
    "            context.user_data[\"worker_detail\"][\"name\"] = worker_name\n",
    "            context.user_data[\"state\"] = \"AddWorkerID\"\n",
    "            reply_markup = buttons[\"ADD_WORKER_NAME_MENU\"]\n",
    "            await update.message.reply_text(f'Got it! Now enter the related username for : {worker_name}', reply_markup=reply_markup)\n",
    "\n",
    "        elif state == \"AddWorkerID\":\n",
    "            worker_name = context.user_data[\"worker_detail\"][\"name\"]\n",
    "            worker_username = update.message.text\n",
    "            context.user_data[\"worker_detail\"][\"username\"] = worker_username\n",
    "            context.user_data['adding_worker'] = False\n",
    "            context.user_data[\"state\"] = None\n",
    "            db_fetcher.add_worker(worker_name, worker_username)\n",
    "            reply_markup = buttons[\"ADD_WORKER_ID_MENU\"]\n",
    "            await update.message.reply_text(f'Worker {worker_name} with username {worker_username} added.', \n",
    "                                           reply_markup=reply_markup)\n",
    "            \n",
    "    elif context.user_data.get('adding_role'):\n",
    "        role_name = update.message.text\n",
    "        context.user_data['adding_role'] = False\n",
    "        await update.message.reply_text(f'Role {role_name} added.')\n",
    "\n",
    "async def add_worker(query: Update, context: ContextTypes.DEFAULT_TYPE) -> None:\n",
    "    \"\"\"Prompt the user to add a worker.\"\"\"\n",
    "    context.user_data['adding_worker'] = True\n",
    "    context.user_data[\"state\"] = \"AddWorkerName\"\n",
    "    await query.edit_message_text('Please enter the name of the worker to add.')\n",
    "\n",
    "async def delete_worker(query, context):\n",
    "    worker_name = context.user_data.get(\"selected-worker\")\n",
    "    db_fetcher.delete_worker(worker_name=worker_name)\n",
    "    await query.edit_message_text(f'Worker {worker_name} deleted')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No error handlers are registered, logging exception.\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_transports\\default.py\", line 69, in map_httpcore_exceptions\n",
      "    yield\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_transports\\default.py\", line 373, in handle_async_request\n",
      "    resp = await self._pool.handle_async_request(req)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpcore\\_async\\connection_pool.py\", line 216, in handle_async_request\n",
      "    raise exc from None\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpcore\\_async\\connection_pool.py\", line 196, in handle_async_request\n",
      "    response = await connection.handle_async_request(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpcore\\_async\\http_proxy.py\", line 317, in handle_async_request\n",
      "    stream = await stream.start_tls(**kwargs)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpcore\\_async\\http11.py\", line 383, in start_tls\n",
      "    return await self._stream.start_tls(ssl_context, server_hostname, timeout)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpcore\\_backends\\anyio.py\", line 68, in start_tls\n",
      "    with map_exceptions(exc_map):\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\contextlib.py\", line 153, in __exit__\n",
      "    self.gen.throw(typ, value, traceback)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpcore\\_exceptions.py\", line 14, in map_exceptions\n",
      "    raise to_exc(exc) from exc\n",
      "httpcore.ConnectError\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\request\\_httpxrequest.py\", line 276, in do_request\n",
      "    res = await self._client.request(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_client.py\", line 1574, in request\n",
      "    return await self.send(request, auth=auth, follow_redirects=follow_redirects)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_client.py\", line 1661, in send\n",
      "    response = await self._send_handling_auth(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_client.py\", line 1689, in _send_handling_auth\n",
      "    response = await self._send_handling_redirects(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_client.py\", line 1726, in _send_handling_redirects\n",
      "    response = await self._send_single_request(request)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_client.py\", line 1763, in _send_single_request\n",
      "    response = await transport.handle_async_request(request)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_transports\\default.py\", line 372, in handle_async_request\n",
      "    with map_httpcore_exceptions():\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\contextlib.py\", line 153, in __exit__\n",
      "    self.gen.throw(typ, value, traceback)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_transports\\default.py\", line 86, in map_httpcore_exceptions\n",
      "    raise mapped_exc(message) from exc\n",
      "httpx.ConnectError\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\ext\\_application.py\", line 1325, in process_update\n",
      "    await coroutine\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\ext\\_handlers\\basehandler.py\", line 157, in handle_update\n",
      "    return await self.callback(update, context)\n",
      "  File \"C:\\Users\\alizeighamiyan\\AppData\\Local\\Temp\\ipykernel_19260\\3631201985.py\", line 45, in button\n",
      "    await query.answer()\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\_callbackquery.py\", line 187, in answer\n",
      "    return await self.get_bot().answer_callback_query(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\ext\\_extbot.py\", line 927, in answer_callback_query\n",
      "    return await super().answer_callback_query(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\_bot.py\", line 3837, in answer_callback_query\n",
      "    return await self._post(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\_bot.py\", line 618, in _post\n",
      "    return await self._do_post(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\ext\\_extbot.py\", line 353, in _do_post\n",
      "    return await super()._do_post(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\_bot.py\", line 647, in _do_post\n",
      "    result = await request.post(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\request\\_baserequest.py\", line 200, in post\n",
      "    result = await self._request_wrapper(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\request\\_baserequest.py\", line 342, in _request_wrapper\n",
      "    raise exc\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\request\\_baserequest.py\", line 332, in _request_wrapper\n",
      "    code, payload = await self.do_request(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\request\\_httpxrequest.py\", line 300, in do_request\n",
      "    raise NetworkError(f\"httpx.{err.__class__.__name__}: {err}\") from err\n",
      "telegram.error.NetworkError: httpx.ConnectError: \n",
      "No error handlers are registered, logging exception.\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_transports\\default.py\", line 69, in map_httpcore_exceptions\n",
      "    yield\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_transports\\default.py\", line 373, in handle_async_request\n",
      "    resp = await self._pool.handle_async_request(req)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpcore\\_async\\connection_pool.py\", line 216, in handle_async_request\n",
      "    raise exc from None\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpcore\\_async\\connection_pool.py\", line 196, in handle_async_request\n",
      "    response = await connection.handle_async_request(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpcore\\_async\\http_proxy.py\", line 317, in handle_async_request\n",
      "    stream = await stream.start_tls(**kwargs)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpcore\\_async\\http11.py\", line 383, in start_tls\n",
      "    return await self._stream.start_tls(ssl_context, server_hostname, timeout)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpcore\\_backends\\anyio.py\", line 68, in start_tls\n",
      "    with map_exceptions(exc_map):\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\contextlib.py\", line 153, in __exit__\n",
      "    self.gen.throw(typ, value, traceback)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpcore\\_exceptions.py\", line 14, in map_exceptions\n",
      "    raise to_exc(exc) from exc\n",
      "httpcore.ConnectError\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\request\\_httpxrequest.py\", line 276, in do_request\n",
      "    res = await self._client.request(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_client.py\", line 1574, in request\n",
      "    return await self.send(request, auth=auth, follow_redirects=follow_redirects)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_client.py\", line 1661, in send\n",
      "    response = await self._send_handling_auth(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_client.py\", line 1689, in _send_handling_auth\n",
      "    response = await self._send_handling_redirects(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_client.py\", line 1726, in _send_handling_redirects\n",
      "    response = await self._send_single_request(request)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_client.py\", line 1763, in _send_single_request\n",
      "    response = await transport.handle_async_request(request)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_transports\\default.py\", line 372, in handle_async_request\n",
      "    with map_httpcore_exceptions():\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\contextlib.py\", line 153, in __exit__\n",
      "    self.gen.throw(typ, value, traceback)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\httpx\\_transports\\default.py\", line 86, in map_httpcore_exceptions\n",
      "    raise mapped_exc(message) from exc\n",
      "httpx.ConnectError\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\ext\\_application.py\", line 1325, in process_update\n",
      "    await coroutine\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\ext\\_handlers\\basehandler.py\", line 157, in handle_update\n",
      "    return await self.callback(update, context)\n",
      "  File \"C:\\Users\\alizeighamiyan\\AppData\\Local\\Temp\\ipykernel_19260\\3631201985.py\", line 39, in start\n",
      "    await update.message.reply_text('Welcome to the Worker Manager Bot! Choose an option:', reply_markup=reply_markup)\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\_message.py\", line 1682, in reply_text\n",
      "    return await self.get_bot().send_message(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\ext\\_extbot.py\", line 2866, in send_message\n",
      "    return await super().send_message(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\_bot.py\", line 1014, in send_message\n",
      "    return await self._send_message(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\ext\\_extbot.py\", line 607, in _send_message\n",
      "    result = await super()._send_message(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\_bot.py\", line 734, in _send_message\n",
      "    result = await self._post(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\_bot.py\", line 618, in _post\n",
      "    return await self._do_post(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\ext\\_extbot.py\", line 353, in _do_post\n",
      "    return await super()._do_post(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\_bot.py\", line 647, in _do_post\n",
      "    result = await request.post(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\request\\_baserequest.py\", line 200, in post\n",
      "    result = await self._request_wrapper(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\request\\_baserequest.py\", line 342, in _request_wrapper\n",
      "    raise exc\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\request\\_baserequest.py\", line 332, in _request_wrapper\n",
      "    code, payload = await self.do_request(\n",
      "  File \"c:\\Users\\alizeighamiyan\\miniconda3\\envs\\digikala\\lib\\site-packages\\telegram\\request\\_httpxrequest.py\", line 300, in do_request\n",
      "    raise NetworkError(f\"httpx.{err.__class__.__name__}: {err}\") from err\n",
      "telegram.error.NetworkError: httpx.ConnectError: \n"
     ]
    }
   ],
   "source": [
    "async def main() -> None:\n",
    "    \"\"\"Start the bot.\"\"\"\n",
    "    # Replace 'YOUR_TOKEN_HERE' with your bot's token\n",
    "\n",
    "    \n",
    "    # Create a new application instance\n",
    "    application = Application.builder().token(TOKEN).build()\n",
    "\n",
    "    # Add handlers to the dispatcher\n",
    "    application.add_handler(CommandHandler(\"start\", start))\n",
    "    application.add_handler(CallbackQueryHandler(button))\n",
    "    application.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, handle_message))\n",
    "\n",
    "    # Run the bot\n",
    "    await application.run_polling()\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    await main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "digikala",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
